airflow/jobs/scheduler_job_runner.py

```diff
@@ -1817,22 +1969,35 @@ def _find_zombies(self) -> None:
         if zombies:
             self.log.warning("Failing (%s) jobs without heartbeat after %s", len(zombies), limit_dttm)
 
-        for ti, file_loc, processor_subdir in zombies:
-            zombie_message_details = self._generate_zombie_message_details(ti)
-            request = TaskCallbackRequest(
-                full_filepath=file_loc,
-                processor_subdir=processor_subdir,
-                simple_task_instance=SimpleTaskInstance.from_ti(ti),
-                msg=str(zombie_message_details),
-            )
-            log_message = (
-                f"Detected zombie job: {request} "
-                "(See https://airflow.apache.org/docs/apache-airflow/"
-                "stable/core-concepts/tasks.html#zombie-undead-tasks)"
-            )
-            self._task_context_logger.error(log_message, ti=ti)
-            self.job.executor.send_callback(request)
-            Stats.incr("zombies_killed", tags={"dag_id": ti.dag_id, "task_id": ti.task_id})
+        with create_session() as session:
+            for ti, file_loc, processor_subdir in zombies:
+                zombie_message_details = self._generate_zombie_message_details(ti)
+                request = TaskCallbackRequest(
+                    full_filepath=file_loc,
+                    processor_subdir=processor_subdir,
+                    simple_task_instance=SimpleTaskInstance.from_ti(ti),
+                    msg=str(zombie_message_details),
+                )
+                session.add(
+                    Log(
+                        event="heartbeat timeout",
+                        task_instance=ti.key,
+                        extra=(
+                            f"Task did not emit heartbeat within time limit ({self._zombie_threshold_secs} "
+                            "seconds) and will be terminated. "
+                            "See https://airflow.apache.org/docs/apache-airflow/"
+                            "stable/core-concepts/tasks.html#zombie-undead-tasks"
+                        ),
+                    )
+                )
+                self.log.error(
+                    "Detected zombie job: %s "
+                    "(See https://airflow.apache.org/docs/apache-airflow/"
+                    "stable/core-concepts/tasks.html#zombie-undead-tasks)",
+                    request,
+                )
+                self.job.executor.send_callback(request)
+                Stats.incr("zombies_killed", tags={"dag_id": ti.dag_id, "task_id": ti.task_id})
```

Comment:
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/rohankokkulabito/airflow/pull/6#issuecomment-2251038846>#720bfc</a><span> - 07/25/2024, 05:35 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The new session context introduced for logging zombie task instances does not handle potential database exceptions. This could lead to unhandled exceptions and potentially leave the session open, causing resource leaks.
</div>
<div id="fix">
<b>Fix:</b> Introduce error handling for the session context to ensure that any database exceptions are properly managed and the session is closed correctly.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -1972,6 +1972,12 @@
          with create_session() as session:
              for ti, file_loc, processor_subdir in zombies:
                  zombie_message_details = self._generate_zombie_message_details(ti)
                  request = TaskCallbackRequest(
                      full_filepath=file_loc,
                      processor_subdir=processor_subdir,
                      simple_task_instance=SimpleTaskInstance.from_ti(ti),
                      msg=str(zombie_message_details),
                  )
                  try:
                      session.add(
                          Log(
                              event="heartbeat timeout",
                              task_instance=ti.key,
                              extra=(
                                  f"Task did not emit heartbeat within time limit ({self._zombie_threshold_secs} "
                                  "seconds) and will be terminated. "
                                  "See https://airflow.apache.org/docs/apache-airflow/"
                                  "stable/core-concepts/tasks.html#zombie-undead-tasks"
                              ),
                          )
                      )
                      session.commit()
                  except Exception as e:
                      self.log.error("Failed to log zombie task instance: %s", e)
                      session.rollback()
                  self.log.error(
                      "Detected zombie job: %s "
                      "(See https://airflow.apache.org/docs/apache-airflow/"
                      "stable/core-concepts/tasks.html#zombie-undead-tasks)",
                      request,
                  )
                  self.job.executor.send_callback(request)
                  Stats.incr("zombies_killed", tags={"dag_id": ti.dag_id, "task_id": ti.task_id})
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged